🐛 Fix Security scopes with string value

Joao-Pedro-P-Holanda opened the pull request:
This pull request fixes a problem when a string value is given to the `scopes` parameter in the `Security` class. Currently, the parameter `scopes` is expected to be a `Sequence[str]`, but python strings fit both the types `str` and `Sequence[str]`, causing every character to be treated as different string, and consequently scope.

e.g.
```python

app = FastAPI()

oauth2_scheme = OAuth2PasswordBearer(
    tokenUrl="token",
    scopes={"me": "Read information about the current user.", "items": "Read items."},
)


def my_auth_function(security_scopes: SecurityScopes, token: Annotated[str, Depends(oauth2_scheme)]):
    return security_scopes.scopes


@app.get("/me")
async def read_single_scope(
    current_scope: Annotated[list[str], Security(my_auth_function, scopes="me")], 
    # type checkers won't give an alert or error in scopes because a str is also a Sequence[str]
):
    return {"scopes": current_scope}
```

Requesting in `/me` with the `Authorization` header set to _"Bearer \<any_text\>"_ would show the scopes as ['m','e'], because the string is treated like a sequence of single characters.

Related: #13368

Joao-Pedro-P-Holanda committed:
test: adding tests to security scopes verification

adding tests to validate an user scope created with a simple string or a
list of strings

Joao-Pedro-P-Holanda committed:
fix: correcting Security scopes when string is given

creating a list explicitly with scopes value when scopes receives a
string

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

Joao-Pedro-P-Holanda committed:
chore: adding type hint for scopes string case

adding Sequence[str] type hint on the self.scopes declaration when
scopes is a string in Security

Joao-Pedro-P-Holanda committed:
fix: removing Annotated usages

removing Annotated to keep compatibility with python 3.8

Joao-Pedro-P-Holanda committed:
fix: correcting redeclaration of function

changing function redeclaration in tests/test_security_oauth2_scopes.py

Kludex commented:
The point here is that `str` is a `Sequence`, so users can't rely on type checkers.

Joao-Pedro-P-Holanda committed:
chore: adding type to get_security_scopes

adding type hint for get_security_scopes in
tests/test_security_oauth2_scopes.py

Joao-Pedro-P-Holanda commented on tests/test_security_oauth2_scopes.py:
```diff
@@ -0,0 +1,49 @@
+)
+
+
+def get_security_scopes(security_scopes: SecurityScopes, token=Depends(oauth2_scheme)):
```
Type hinting as `list[str]`  doesn't work for python 3.8 :cry:, i also found that almost all test files don't use type hintings.

One possiblity is to hint with just `Sequence` or `list`, but now type checkers would complain that "type arguments expected for generic class".

What do you think about?  Use just `Sequence` or `list` or don't put type hinting at all for this function?

alv2017 replied:
In this particular place we need to show our intentions and be explicit :smiley: 

```
from typing import List
```
should work for python 3.8+

alv2017 replied:
By the way this is a very good point, Python 3.8 is not supported any longer https://devguide.python.org/versions/

alv2017 commented on tests/test_security_oauth2_scopes.py:
```diff
@@ -0,0 +1,51 @@
+
+oauth2_scheme = OAuth2PasswordBearer(
+    tokenUrl="token",
+    scopes={"me": "Read information about the current user.", "items": "Read items."},
```
Scope definition syntax: https://datatracker.ietf.org/doc/html/rfc6749#section-3.3
(I express doubts regarding the correctness of the `scopes`)

Joao-Pedro-P-Holanda committed:
fix: type hinting get_security_scopes with typing List

using List from typing package, as python's list can't be subscriptable
in python 3.8

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

alv2017 commented:
Be careful with types !!! 
Reminder: https://fastapi.tiangolo.com/reference/security/#fastapi.security.SecurityScopes

Joao-Pedro-P-Holanda commented:
@alv2017 could you be clearer? The type of the scopes instance attribute is also `List[str]` in `SecurityScopes`

https://github.com/fastapi/fastapi/blob/eea196f4a505c2584fafda691f2a0033c4db6622/fastapi/security/oauth2.py#L622-L628

alv2017 commented:
@Joao-Pedro-P-Holanda 
1) The source code was written quite sometime ago, hence it uses the deprecated annotation typing.List. I think that at the moment you shouldn't worry about it.

2) However in this PR you are introducing a new feature, and it is going to affect the source code and the docs (including the Security API). So it is important to be careful with that.

Joao-Pedro-P-Holanda committed:
docs: documenting edge case in Security's scopes parameter

adding a short information about the case where a string is given in
place of a list of strings (or similar) on the Doc of the Security
function in fastapi/param_functions.py

Joao-Pedro-P-Holanda commented:
Now i understood, thx. 
In my opinion, only the references documentation needs an update based on this change. As the behavior for all other cases, like in https://fastapi.tiangolo.com/advanced/security/oauth2-scopes/, remain the same.

https://fastapi.tiangolo.com/reference/dependencies/#fastapi.Security

If the documentation update is not needed or a different file needs update i can revert [5e2896f](https://github.com/fastapi/fastapi/pull/13369/commits/5e2896fd3c0b0aa880a9daec873c02e595da18b7)

svlandeg committed:
Merge branch 'master' into fix/security-string-scopes

Joao-Pedro-P-Holanda committed:
Merge branch 'master' into fix/security-string-scopes

Joao-Pedro-P-Holanda committed:
Merge branch 'master' into fix/security-string-scopes

YuriiMotov reviewed (commented):
I think the type of input parameter should be changed from
```python
scopes: Optional[Sequence[str]] = None
```

to 
```python
scopes: Optional[Union[str, Sequence[str]]] = None
```

Here:
```python
class Security(Depends):
    def __init__(
        self,
        dependency: Optional[Callable[..., Any]] = None,
        *,
        scopes: Optional[Sequence[str]] = None,
        use_cache: bool = True,
    ):
```

Joao-Pedro-P-Holanda committed:
refactor: change Security class typing for scopes param

Make scopes a str or Sequence of str explicitly

Joao-Pedro-P-Holanda committed:
fix: use Union instead of pipe in scopes type hint

YuriiMotov reviewed (dismissed):
LGTM

Joao-Pedro-P-Holanda committed:
Merge remote-tracking branch 'upstream/master' into fix/security-string-scopes
